server/resp/src/main/java/org/infinispan/server/resp/commands/list/LMOVE.java

```diff
@@ -0,0 +1,154 @@
+      }
+
+      return lmoveAndReturn(handler, ctx, arguments, false);
+   }
+
+   protected CompletionStage<RespRequestHandler> lmoveAndReturn(Resp3Handler handler,
+                                                                ChannelHandlerContext ctx,
+                                                                List<byte[]> arguments,
+                                                                boolean rightLeft) {
+      byte[] source = arguments.get(0);
+      byte[] destination = arguments.get(1);
+
+      boolean txCache = handler.cache().getCacheConfiguration().transaction().transactionMode().isTransactional();
+      boolean sameList = source.equals(destination);
+
+      if (!sameList && !txCache) {
+         // different lists and not tx cache, warn
+         Log.SERVER.lmoveConsistencyMessage();
+      }
+
+      final boolean isSourceLeft;
+      final boolean isDestinationLeft;
+
+      if (rightLeft) {
+         isSourceLeft = false;
+         isDestinationLeft = true;
+      } else {
+         // parse and validate RIGHT and LEFT arguments
+         final String sourceWhereFrom = new String(arguments.get(2)).toUpperCase();
+         final String destinationWhereFrom = new String(arguments.get(3)).toUpperCase();
+         isSourceLeft = LEFT.equals(sourceWhereFrom);
+         isDestinationLeft = LEFT.equals(destinationWhereFrom);
+         if ((!isSourceLeft && !RIGHT.equals(sourceWhereFrom)) || (!isDestinationLeft && !RIGHT.equals(
+               destinationWhereFrom))) {
+            RespErrorUtil.syntaxError(handler.allocator());
+            return handler.myStage();
+         }
+      }
+
+      EmbeddedMultimapListCache<byte[], byte[]> listMultimap = handler.getListMultimap();
+      if (sameList) {
+         CompletionStage<byte[]> performedCall;
+         if (isSourceLeft && isDestinationLeft) {
+            // peek first element and do nothing else
+            performedCall = listMultimap.index(source, 0);
+         } else if (!isSourceLeft && !isDestinationLeft) {
+            // peek last element and do nothing else
+            performedCall = listMultimap.index(source, -1);
+         } else {
+            // rotate from left (head->tail) to right or from right to left (tail->left)
+            performedCall = listMultimap.rotate(source, isSourceLeft);
+         }
+         return CompletionStages.handleAndCompose(performedCall, (result, t) -> {
+            if (t != null) {
+               return handleException(handler, t);
+            }
+
+            return handler.stageToReturn(CompletableFuture.completedFuture(result), ctx, Consumers.GET_BICONSUMER);
+         });
+      }
+
+      CompletionStage<Collection<byte[]>> pollCall;
+      if (isSourceLeft) {
+         pollCall = listMultimap.pollFirst(source, 1);
+      } else {
+         pollCall = listMultimap.pollLast(source, 1);
+      }
+
+      return CompletionStages.handleAndCompose(pollCall, (pollResult, t1) -> {
```

Comment:
If we get #11038 first, we can simplify the nested `handleAndCompose` and use only the `thenCompose` instead.